examples/audio/soundtrack.rs

```diff
@@ -0,0 +1,157 @@
+
+// This component will be attached to an entity to fade the audio out
+#[derive(Component)]
+struct FadeOut;
+
+fn setup(asset_server: Res<AssetServer>, mut commands: Commands) {
+    // Instantiate the game state resources
+    commands.insert_resource(GameState::default());
+    commands.insert_resource(GameStateTimer(Timer::from_seconds(
+        10.0,
+        TimerMode::Repeating,
+    )));
+
+    // Create the track list
+    let track_1 = asset_server.load::<AudioSource>("sounds/Mysterious acoustic guitar.ogg");
+    let track_2 = asset_server.load::<AudioSource>("sounds/Epic orchestra music.ogg");
+    let track_list = vec![track_1, track_2];
+    commands.insert_resource(SoundtrackPlayer::new(track_list));
+}
+
+// Every time the GameState resource changes, this system is run to trigger the song change.
+fn change_track(
+    mut commands: Commands,
+    soundtrack_player: Res<SoundtrackPlayer>,
+    soundtrack: Query<Entity, With<AudioSink>>,
+    game_state: Res<GameState>,
+) {
+    if game_state.is_changed() {
+        // Fade out all currently running tracks
+        for track in soundtrack.iter() {
+            commands.entity(track).insert(FadeOut);
+        }
+
+        // Spawn a new `AudioBundle` with the appropriate soundtrack based on
+        // the game state.
+        //
+        // Volume is set to start at zero and is then increased by the fade_in system.
+        match game_state.as_ref() {
+            GameState::Peaceful => {
+                commands.spawn((
+                    AudioBundle {
+                        source: soundtrack_player.track_list.first().unwrap().clone(),
+                        settings: PlaybackSettings {
+                            mode: bevy::audio::PlaybackMode::Loop,
+                            volume: bevy::audio::Volume::ZERO,
+                            ..default()
+                        },
+                    },
+                    FadeIn,
+                ));
+            }
+            GameState::Battle => {
+                commands.spawn((
+                    AudioBundle {
+                        source: soundtrack_player.track_list.get(1).unwrap().clone(),
+                        settings: PlaybackSettings {
+                            mode: bevy::audio::PlaybackMode::Loop,
+                            volume: bevy::audio::Volume::ZERO,
+                            ..default()
+                        },
+                    },
+                    FadeIn,
+                ));
+            }
+        }
+    }
+}
+
+// Fade effect duration
+const FADE_TIME: f32 = 2.0;
+
+// Fades in the audio of entities that has the FadeIn component. Removes the FadeIn component once
+// full volume is reached.
+fn fade_in(
+    mut commands: Commands,
+    mut audio_sink: Query<(&mut AudioSink, Entity), With<FadeIn>>,
+    time: Res<Time>,
+) {
+    for (audio, entity) in audio_sink.iter_mut() {
+        audio.set_volume(audio.volume() + time.delta_seconds() / FADE_TIME);
```

Comment:
Nothing we can do at this stage for the example but this highlights the need for tweening in the audio engine -- this is going to result in sounding like stepwise motion, because it will only update as fast as the game renders (which is 60 or maybe even 120 Hz which is definitely not fast enough to perceive as gradual).

In this case because the fade is 2 seconds each increment is going to be gradual enough, but shorten the fade time and the problem manifests more clearly.

Reply:
This is also something I considered while programming this. Also, I just did a linear fade on the volume setting, but according to the docs, to decrease the perceived volume by half, you should multiply by 0.316. I figured it was good enough at illustrating the concept while still keeping things simple.